runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,449 @@
+package org.corfudb.runtime.collections;
+
+import static org.corfudb.runtime.collections.QueryOptions.DEFAULT_OPTIONS;
+
+import com.google.protobuf.Message;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.TreeSet;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.runtime.CorfuStoreMetadata.Timestamp;
+import org.corfudb.runtime.object.transactions.Transaction.TransactionBuilder;
+import org.corfudb.runtime.object.transactions.TransactionType;
+import org.corfudb.runtime.view.ObjectsView;
+import org.corfudb.runtime.view.TableRegistry;
+
+/**
+ * Created by zlokhandwala on 2019-08-09.
+ */
+public class Query {
+
+    private final TableRegistry tableRegistry;
+
+    private final ObjectsView objectsView;
+
+    private final String namespace;
+
+    /**
+     * Creates a Query interface.
+     *
+     * @param tableRegistry Table registry from the corfu client.
+     * @param objectsView   Objects View from the corfu client.
+     * @param namespace     Namespace to perform the queries within.
+     */
+    public Query(final TableRegistry tableRegistry, final ObjectsView objectsView, final String namespace) {
+        this.tableRegistry = tableRegistry;
+        this.objectsView = objectsView;
+        this.namespace = namespace;
+    }
+
+    private Token getToken(Timestamp timestamp) {
+        return new Token(timestamp.getEpoch(), timestamp.getSequence());
+    }
+
+    private void txBegin(Timestamp timestamp) {
+        TransactionBuilder transactionBuilder = objectsView.TXBuild()
+                .type(TransactionType.SNAPSHOT);
+
+        if (timestamp != null) {
+            transactionBuilder.snapshot(getToken(timestamp));
+        }
+
+        transactionBuilder.build().begin();
+    }
+
+    private void txEnd() {
+        objectsView.TXEnd();
+    }
+
+    private <K extends Message, V extends Message, M extends Message>
+    Table<K, V, M> getTable(@Nonnull final String tableName) {
+        return tableRegistry.getTable(this.namespace, tableName);
+    }
+
+    /**
+     * Fetch the Value payload for the specified key.
+     *
+     * @param tableName Table name.
+     * @param key       Key.
+     * @param <K>       Type of key.
+     * @param <V>       Type of value.
+     * @return Value.
+     */
+    @Nullable
+    public <K extends Message, V extends Message, M extends Message>
+    CorfuRecord<V, M> get(@Nonnull final String tableName,
```

Comment:
Should this return value instead of CorfuRecord? This currently looks exactly same as getRecord().

Reply:
sorry for the confusion but the purpose of overloading this get() method is to have one with timestamp specified and one without. Note below there is a get() where caller can explicitly specify a timestamp to run the query on. 
I have updated the comment to reflect the same. 
thanks

Reply:
got rid of these extra get() and just retained getRecord